doc/source/templates/testing/cluster_envs/04_finetuning_llms_with_deepspeed.yaml

```diff
@@ -28,6 +28,8 @@ python:
   conda_packages: []
 
 post_build_cmds:
+  - pip uninstall bitsandbytes -y || true
+  - pip install torch==2.1.1 --index-url https://download.pytorch.org/whl/cu118
```

Comment:
These changes just make sure that the testing environment is as close to the template as possible while using nightly.

Reply:
why can't we add torch version as part of the `pip_packages`?

Reply:
isn't this also installing the cu118 version?

Reply:
hmm, is torch 2.1.1 actually needed? cu118 supports H100 apparently: https://deeptalk.lambdalabs.com/t/how-to-run-h100-with-docker-in-sm-86-compatibility-mode/3730/2

and torch 2.1+ is only needed to support cu121

Reply:
https://github.com/pytorch/pytorch/issues/90761#issuecomment-1673709633
I was going from this comment from Patrick and installed 2.1.1 since it supports CUDA 12.1.
CUDA 12.1 is the CUDA version that we use in our Dockerfile for this template `doc/source/templates/testing/docker/04_finetuning_llms_with_deepspeed/Dockerfile`.
Therefore, I through I'd just go with that CUDA version and see if everything works.
I've since built the Dockerfile and tested it in a workspace and everything is working fine.

Reply:
> isn't this also installing the cu118 version, not cu121?

Yes, that's correct. But the nightly image we use for testing only has CUDA 11.8.
I didn't want to mess with how we test this so I thought I might just as well use CUDA 11.8.
After all, as you said, CUDA 11.8 should suffice to support H100s.

Reply:
> why can't we add torch version as part of the `pip_packages`?

I tried that but pip got somehow thrown off by the `--index-url https://download.pytorch.org/whl/cu118`. I attempted to adhere to how it's supposed to go into a pipfile like
```
....
--index-url https://download.pytorch.org/whl/cu118
torch==2.1.1 
....
```
But pip ended up attempting to resolve the other deps from the pytorch registry. So I stopped trying it the pretty way and this is the product of that process.

Reply:
i see. so the only thing is to make sure that none of the other packages have `torch` as a dependency of a lower version. in that case, `pip install torch==2.1.1` will noop. you can do `pip install -U torch==2.1.1` to make sure it gets upgraded, but that also means that it may be incompatible with other packages, and may even update other packages to different versions than in the `requirements.txt`

Reply:
Yes. What solution would you propose?

Reply:
@justinvyu gentle ping

Reply:
Oh. idk I was thinking we can try to figure out the requirements.txt file issue 😆 But I am ok with this, I was just noting it down as a warning.